### PR TITLE
TxnId should be more succinct and more useful

### PR DESCRIPTION
**Pull request by belliottsmith**
Specifically, TxnId should have no more than 20 bytes of payload, and should encode simple flags about a Txn, such as whether it is a read or write transaction, and whether it operates on keys or ranges.

### CODE REVIEW

File: accord-core/src/main/java/accord/primitives/RangeDeps.java [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package accord.primitives;
```
Comment (iamaleksey) [inline accord-core/src/main/java/accord/primitives/RangeDeps.java]: Stray from adjacent work?

File: accord-core/src/main/java/accord/impl/InMemoryCommandsForKey.java [hunk 2]
```diff
@@ -37,7 +37,7 @@
 import static accord.local.CommandsForKey.CommandTimeseries.TestKind.RorWs;
 import static accord.local.Status.KnownDeps.DepsUnknown;
 import static accord.local.Status.PreAccepted;
-import static accord.primitives.Txn.Kind.WRITE;
+import static accord.primitives.Txn.Kind.Write;
```
Comment (iamaleksey) [inline accord-core/src/main/java/accord/impl/InMemoryCommandsForKey.java]: Dead import

File: accord-core/src/main/java/accord/messages/BeginInvalidation.java [hunk 3]
```diff
@@ -35,7 +35,7 @@ public BeginInvalidation(TxnId txnId, Unseekables<?, ?> someUnseekables, Ballot
```
Comment (iamaleksey) [inline accord-core/src/main/java/accord/messages/BeginInvalidation.java:35]: Pre-existing: missing `@Override`

File: accord-core/src/main/java/accord/messages/InformOfTxnId.java [hunk 4]
```diff
@@ -25,7 +23,7 @@ public InformOfTxnId(TxnId txnId, RoutingKey homeKey)
     public void process()
```
Comment (iamaleksey) [inline accord-core/src/main/java/accord/messages/InformOfTxnId.java:24]: Pre-existing: missing `@Override`

File: accord-core/src/test/java/accord/txn/DepsTest.java [hunk 5]
```diff
@@ -312,21 +312,21 @@ static Deps generate(Gen.Random random)
         {
             int epochRange = 3;
             int realRange = 50;
```
Comment (iamaleksey) [inline accord-core/src/test/java/accord/txn/DepsTest.java]: Rename to hlcRange?

File: accord-core/src/main/java/accord/primitives/Ballot.java [hunk 6]
```diff
@@ -33,4 +38,9 @@ public Ballot(long epoch, long real, int logical, Id node)
     {
```
Comment (iamaleksey) [inline accord-core/src/main/java/accord/primitives/Ballot.java:48]: This constructor is a dirty liar.

File: accord-core/src/test/java/accord/coordinate/TopologyChangeTest.java [hunk 7]
```diff
@@ -35,20 +32,18 @@
...
 import static accord.Utils.*;
 import static accord.impl.IntKey.keys;
 import static accord.impl.IntKey.range;
 import static accord.local.PreLoadContext.empty;
+import static accord.primitives.Routable.Domain.Key;
+import static accord.primitives.Txn.Kind.Write;
 
 public class TopologyChangeTest
 {
     private static TxnId coordinate(Node node, Keys keys) throws Throwable
```
Comment (iamaleksey) [inline accord-core/src/test/java/accord/coordinate/TopologyChangeTest.java]: Unused method?

File: accord-core/src/main/java/accord/primitives/TxnId.java [hunk 8]
```diff
@@ -22,13 +22,42 @@
 
 public class TxnId extends Timestamp
 {
-    public TxnId(Timestamp timestamp)
+    public static TxnId fromBits(long msb, long lsb, Id node)
     {
-        super(timestamp);
+        return new TxnId(msb, lsb, node);
+    }
```
Comment (iamaleksey) [inline accord-core/src/main/java/accord/primitives/TxnId.java:33]: Nit: newline between methods?

File: accord-core/src/main/java/accord/local/Command.java [hunk 9]
```diff
@@ -295,12 +292,12 @@ public AcceptOutcome accept(SafeCommandStore safeStore, Ballot ballot, Kind kind
             return AcceptOutcome.Redundant;
         }
 
-        if (known().isDefinitionKnown() && !kind().equals(kind))
+        if (known().isDefinitionKnown() && !txnId().rw().equals(kind))
```
Comment (iamaleksey) [inline accord-core/src/main/java/accord/local/Command.java]: ==?

File: accord-core/src/main/java/accord/local/Command.java [hunk 10]
```diff
@@ -977,7 +973,7 @@ else if (existingRanges != additionalRanges && !route().covers(additionalRanges)
         if (!validate(ensurePartialTxn, existingRanges, additionalRanges, covers(partialTxn()), covers(partialTxn), "txn", partialTxn))
             return false;
 
-        if (partialTxn != null && kind() != null && !kind().equals(partialTxn.kind()))
+        if (partialTxn != null && !txnId().rw().equals(partialTxn.kind()))
```
Comment (iamaleksey) [inline accord-core/src/main/java/accord/local/Command.java]: ==?

File: accord-core/src/main/java/accord/primitives/TxnId.java [hunk 11]
```diff
@@ -22,13 +22,42 @@
...
+    {
+        super(epoch, hlc, flags, node);
+    }
+
+    private TxnId(long msb, long lsb, Id node)
+    {
+        super(msb, lsb, node);
+    }
+
+    public Txn.Kind rw()
```
Comment (iamaleksey) [inline accord-core/src/main/java/accord/primitives/TxnId.java]: I would consider adding `isWrite()` and `isRead()` to `TxnId` to further shorten call chains like `txnId.rw().isWrite()` in e.g. `InMemoryCommandsForKey` by one.

File: accord-core/src/main/java/accord/primitives/Ballot.java [hunk 12]
```diff
@@ -22,15 +22,35 @@
...
         super(from);
     }
 
-    public Ballot(long epoch, long real, int logical, Id node)
+    Ballot(long epoch, long hlc, int flags, Id node)
+    {
+        super(epoch, hlc, flags, node);
+    }
+
+    private Ballot(long msb, long lsb, Id node)
```
Comment (iamaleksey) [inline accord-core/src/main/java/accord/primitives/Ballot.java]: OCD: give both bottom two constructors same visibility? public, private, or package-private, doesn't matter.

File: accord-core/src/main/java/accord/primitives/Timestamp.java [hunk 13]
```diff
@@ -19,62 +19,126 @@
...
-        return new Timestamp(epoch, real, logical + 1, node);
+        long lsb = this.lsb + HLC_INCR;
+        long msb = this.msb;
+        if (Long.compareUnsigned(this.lsb, lsb) > 0)
+            ++msb; // overflow of lsb
+        return new Timestamp(msb, lsb, node);
     }
 
     @Override
     public int compareTo(Timestamp that)
```
Comment (iamaleksey) [inline accord-core/src/main/java/accord/primitives/Timestamp.java]: `@NotNull that` to makе IDEA happy?

File: accord-core/src/main/java/accord/primitives/Timestamp.java [hunk 14]
```diff
@@ -19,62 +19,126 @@
...
+    {
+        return (msb << 48) | (lsb >>> 16);
+    }
+
+    public int flags()
+    {
+        return (int) (lsb & MAX_FLAGS);
+    }
+
     public Timestamp withMinEpoch(long minEpoch)
```
Comment (iamaleksey) [inline accord-core/src/main/java/accord/primitives/Timestamp.java]: Maybe rename to `withEpochAtLeast()` or something along those lines? My initial understanding was: with min( epoch) out of {current, provided}, but it does the opposite of course.

File: accord-core/src/main/java/accord/primitives/Timestamp.java [hunk 15]
```diff
@@ -19,62 +19,126 @@
...
 
     @Override
     public int compareTo(Timestamp that)
     {
         if (this == that) return 0;
-        int c = Long.compare(this.epoch, that.epoch);
-        if (c == 0) c = Long.compare(this.real, that.real);
-        if (c == 0) c = Integer.compare(this.logical, that.logical);
+        int c = Long.compareUnsigned(this.msb, that.msb);
+        if (c == 0) c = Long.compareUnsigned(this.lsb, that.lsb);
```
Comment (iamaleksey) [inline accord-core/src/main/java/accord/primitives/Timestamp.java:139]: Arbitrary as it is, don't we want to compare `Node.Id` before `flags`?

Comment (iamaleksey) [reply accord-core/src/main/java/accord/primitives/Timestamp.java:139]: To allow byte-order comparability (if ever relevant), when `Node.Id` becomes `int`, can spread it between `lsb` and the extra int, moving `flags` to the end of the logical byte-sequence.

**Review by iamaleksey — CHANGES_REQUESTED** [review]
Basically LGTM. Some unresolved suggestions in Conversation section still. All minor, none binding, except for maybe enum comparisons using `equals()`.

File: accord-core/src/main/java/accord/primitives/Timestamp.java (hunk 15 above)
Comment (belliottsmith) [reply accord-core/src/main/java/accord/primitives/Timestamp.java:139]: I've opted to ignore the flag bits entirely for identity purposes, to permit us to use them as carrier bits for things that do not relate 1:1 with the identity bits. We can move their position around later.

**Comment by belliottsmith** [comment]
Thanks for the swift review. I think I've integrated every suggestion, but let me know if I missed anything.

File: accord-core/src/main/java/accord/primitives/Timestamp.java [hunk 16]
```diff
@@ -19,62 +19,145 @@
...
+    }
+
+    public static Timestamp fromValues(long epoch, long hlc, int flags, Id node)
+    {
+        return new Timestamp(epoch, hlc, flags, node);
+    }
+
+    public static Timestamp maxForEpoch(long epoch)
+    {
+        return new Timestamp((epoch << 16) | 0xffff, Long.MAX_VALUE, Id.MAX);
```
Comment (iamaleksey) [inline accord-core/src/main/java/accord/primitives/Timestamp.java]: `epochMsb(epoch)` for consistency?

File: accord-core/src/main/java/accord/primitives/Timestamp.java [hunk 17]
```diff
@@ -19,62 +19,145 @@
...
+    }
+
+    public static Timestamp minForEpoch(long epoch)
+    {
+        return new Timestamp(epoch << 16, 0, Id.NONE);
```
Comment (iamaleksey) [inline accord-core/src/main/java/accord/primitives/Timestamp.java]: `epochMsb(epoch)` for consistency?

File: accord-core/src/main/java/accord/primitives/Timestamp.java (hunk 16 above)
Comment (iamaleksey) [reply accord-core/src/main/java/accord/primitives/Timestamp.java:45]: Also, `MAX_FLAGS`?

**Review by iamaleksey — APPROVED** [review]
LGTM

**Pull request merged**